Python template support

radumg opened the pull request:
### Purpose

This PR fixes #7604 , wish for adding a Python template that can be used to populate any `Python Script` nodes when adding to workspace.

#### Current behaviour
- user adds `Python Script` node to canvas
- the node is populated with a few lines of code, but these are hard-coded in the Dynamo source code, so users are unable to change them

#### New behaviour
- user adds `Python Script` node to canvas
- Dynamo checks if the `PythonTemplate.py` file exists at the user location root (`%appdata%/Dynamo/Core/{version}/`) and if it does, it reads the file and populates the Python script node with its contents
- if the file doesn't exits or it's empty, Dynamo falls back on the hard-coded behaviour

![pythontemplate](https://user-images.githubusercontent.com/11439624/28098410-095e2f70-66ae-11e7-8a24-2833a4563921.gif)

This allows users and organisations to rollout the template as part of users profiles or as logon scripts.

### Implementation
- refactored `PreferenceSettings` class to include code regions, moving properties into their logical group, also fixing some comments/documentation at same time, making it clearer.
- implemented a `PythonTemplateFilePath` property in the settings class so it serialises and deserialises with user settings
- implemented required properties in the interfaces `IPreferences`, `IPathResolver` and `PathManager` class
- added a few additional lines to the hardcoded python template, providing pointers to users where to start writing code and explaining the first lines. this is following my observations in-person and on forums that most new users struggle with those 2 things : where to put code and what those bits mean.
- replaced `\n` with .NET standard `Environment.NewLine` line breaks, should behave more predictably when compiled on different platforms

#### Limitations
- the name and location of template file is hardcoded. however, this is consistent/same as `DynamoSettings.XML` file itself. Since the interfaces implement the property, it can be exposed via UI in later PR
![path](https://user-images.githubusercontent.com/11439624/28098400-f658cb88-66ad-11e7-8641-b162ae4f7c95.PNG)
- I couldn't figure out how to access the current `DynamoModel` instance from the `PythonNodeModel` class to retrieve current settings, so the template is read every time a `Python Script` node is added to canvas. Would welcome some pointers here as both the Dynamo API documentation and the class annotations are lackluster. Would've expected either a static class or some public methods that expose this.

### Declarations
- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers
@kronz 
@mjkkirschner
@QilongTang  

### FYIs
@ksobon , @ThomasMahon, @wynged , @eibre , @andydandy74

brencass commented:
Sterling work and this is a great addition.

I have a small command; Instead of the location being hard coded as being in a user location, could this location not be set from within the Dynamosettings.xml file, but if not filled in then it goes to default location?

This will allow companies to have a central location for company standards that includes this python template and dynamo template.

radumg commented:
Thanks @brencass !

Really good suggestion - that's what i had in mind too and this PR does indeed store and read the location in the `DynamoSettings.XML` file, so that bit is already working.
What I couldn't do is access that setting from within the code that initialises the `Python Script` node - it was already 3am by that point though, so i might have been a bit blind :grin: 

If someone a bit more familiar with the codebase ( @ke-yu, @QilongTang , @mjkkirschner )  can weight in and point me in the right direction, i can connect the 2 dots as originally intended.

radumg commented:
Ok, so I implemented requested change by @brencass above and aligned it to what I originally envisioned.

Added a `public`  `static` property called `PublicPreferenceSettings` to `DynamoModel` class, so that nodes and other components that are not fully aware of the context they're executed in can still access these properties.

The property should be thread-safe as it's read-only :
- read : `public` access
- write : `private` access

Also added lines to update the static property whenever the instance-based `PreferenceSettings` gets updated within the `DynamoModel` class.

Tested by

- [x] building entire solution
- [x] running `DynamoSandbox`
- [x] changing Dynamo settings (such as `show connectors` or adding a new custom package path)
- [x] nodes reading the static property see the updates

Notes :
- you can't (from what I can tell and what @mjkkirschner [hinted](https://twitter.com/MJKKirschner/status/885628144090308608) at) get the current `DynamoModel` context from a node such as the Python Node
- this property duplicates the contents of `PreferenceSettings` property, so any time this prop is updated, the static one is also updated
- this duplication is rather inelegant but I don't see another way (rebuilding the settings file path and reading + parsing XML file for each Python node added to canvas is worse I think as it introduces duplicate code)

radumg commented:
@Racel , @ikeough  - could you please have a look and let me know if this is something you'd want merged in and if any further changes are required ?

Racel commented:
@radumg - We have a meeting to review this today. Will discuss with the team and get back to you.

Racel commented:
@radumg - After review with the team, we want to consider other approaches than changing DynamoModel. @mjkkirschner will follow-up with some suggestions to think about.

Additionally, we noticed that your PRs were made against the 1.3.1 Release Candidate. In order for your changes to appear in the next stable release of Dynamo, please make a PR against master.

radumg commented:
Thank you @Racel for the feedback :+1: and quick reply!
Looking forward to hearing @mjkkirschner 's suggestions - I agree changing the `DynamoModel` isn't ideal so happy to make changes.

As for the PR branches, my bad : I assumed `1.3.1` would be the next stable release since there wasn't anything published after `1.3`. It's since become apparent it was a  `Revit 2018.1`-only release so I'll rebase on `master` branch and deal with any conflicts.

radumg commented:
@mjkkirschner , in anticipation of your suggestions : my initial instinct was to turn the PreferenceSettings class to `static` and refactor references to it as needed.
This however introduces the question of how this will be handled when support for multiple workspaces is added in 2.0 : my view is since they're global settings, they don't need to be in the `DynamoModel` in the first place, as they should not vary across models/workspaces.

This is obviously a significant departure from current approach & also a UX question, so i await instructions on what you think will be best.
Also, this has been a really good learning opportunity for me, thanks to the team for taking the time to review 👍

mjkkirschner commented:
@radumg so here are some alternatives I have not given much thought 😉 .

* from python node raise an event requesting the path, dynamoModel can handle this and send back the path - it could be a general mechanism for requesting any preference setting even. This might be added to the DynamoServices project.

* a static property on the PreferencesSettings class instead of DynamoModel.

* a static property added to Dynamo services.

* Python Node has a customization so it would be possible to find the dynamoModel through the DynamoViewModel, but I would advise against that, it means the execution would be tied to the view layer.

In general, I think adding more properties to DynamoModel is not great as it's already a giant class, so I'd rather if we have to that we add it somewhere specific.

let us know what you think of those.

radumg commented:
hi @mjkkirschner , thanks for all the options above! And sorry for the delay in replying, we've been busy organising the latest UK Dynamo User Group meeting last week.

## Options feedback
In terms of the options, my notes on them would be : 

**from python node raise an event**
```
This could be a good option to generally fetch properties from within nodes, the drawback I see would be :
- potential performance issues in waiting for the event callback when adding a new node ? (I don't know what order the event requests are processed in so am wondering if it's possible the UI layer would be locked waiting for this callback)
- introduces additional code in each node/codebase requiring access to these properties
- I might have a follow-up question or 2 about this one :D
```

**a static property on the PreferencesSettings class instead of DynamoModel**
```quickest one to implement - happy to go ahead with this one.```

**a static property added to Dynamo services**
``` the only place I can find that's vaguely suitable to place the static property would be the `IExecutionSession` interface - suggesting to me that `DynamoServices` might not be the right or expected place to implement this after all, since it would still require adding the property to the `DynamoModel` class which implements that interface.```

**Python Node customization**
``` agree ```

## Suggested options

That would leave us I think with 3 main options :

1. a static property on the PreferencesSettings class
2. a static `PreferencesSettings` class
3. event handling on `DynamoServices` project

I'd gravitate towards `option 2` above as a more long-term solution, I'm happy to go through implications and resolve. The lens I see this through is that once this PR is resolved, i'll be working on `DYN template` support next.
Still requires the answer to my question above ```When support for multiple workspaces is added, does each DynamoModel still require their own Preferences ?``` Could simplify the DynamoModel class too.

Let me know which option you want me to go for,
```1 - static property```
```2 - static Preferences class```
```3 - event handling in DynamoServices```

Cheers, Radu

mjkkirschner commented:
@radumg sorry for the delay here.
I think there might be someway to adding this info to dynamoServices but as I looked through it not with any of the constructs we currently have there as they are all event based... let's do option number 2 for now and add a static property specifically for the`pythonTemplatePath` to the preferenceSettings class. This will let us control what preferences are made accessible to other nodes deliberately. Sound good?

mjkkirschner commented:
@radumg to clarify: 
nubmer 1 in your final list :).
`1 - static property in PreferenceSettings class`

mjkkirschner commented on src/DynamoCore/Configuration/IPathResolver.cs:
```diff
@@ -158,6 +158,12 @@ public interface IPathManager
+        /// Full path to the python template py file. This setting file is specific 
+        /// to the current user.
+        /// </summary>
+        string PythonTemplateFilePath { get; }
```
because you are changing the interface here you may need to update: https://github.com/DynamoDS/DynamoRevit/blob/Revit2017/src/DynamoRevit/RevitPathResolver.cs

in a related PR to avoid breaking the Revit build.

radumg replied:
Thanks @mjkkirschner , was definitely not aware of Dynamo Revit was implementing its own class from this interface. Will try and keep an eye out for that in the future.

mjkkirschner commented on src/DynamoCore/Configuration/PreferenceSettings.cs:
```diff
@@ -304,6 +322,8 @@ public PreferenceSettings()
             ShowDetailedLayout = true;
             NamespacesToExcludeFromLibrary = new List<string>();
 
+            PythonTemplateFilePath = new PathManager(new PathManagerParams()).PythonTemplateFilePath;
```
I've not investigated far enough to know, but do you actually need to create a new path manager just to pull this property out, don't we create a pathmanager somewhere else during dynamoModel initialization?

radumg replied:
As far as I remember, when I wrote this I couldn't find a way to access any part of the current Dynamo Model instance.
You are right, a `PathManager` is instantiated as part of the `PreferenceSettings` class, which is created when making a `DynamoModel`, but since I couldn't get to it, resorted to the work-around of making a new dummy instance of `PathManager`. The static property will fix that, so this line will no longer be necessary.

mjkkirschner commented on src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs:
```diff
@@ -45,7 +45,7 @@
 // in AssemblyVersionInfo.cs so that it can be easily incremented by the
 // automated build process.
-[assembly: AssemblyVersion("1.3.1.1514")]
+[assembly: AssemblyVersion("1.3.1.1977")]
```
I think you should send future PRs to master - and no need to commit this file, its auto generated.

radumg replied:
Will do, I will submit a new PR based on `master` for this one too. I originally based this one off the `1.3.1` branch thinking that would be the next release (it wasn't out yet). Maybe it would be a good idea to add to the [wiki](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md) a bit about git branch management and how the team uses them, which one is current/release/etc ? I remember first PR I did was highly confusing how it was cherry-picked when merged etc.

radumg replied:
Dug around...can't figure out why this file was included, it's in the [`.gitignore` ](https://github.com/DynamoDS/Dynamo/blob/master/.gitignore#L74) for the repo, which is what my cloned version uses too.
Thanks for catching it, I'll exclude from future PRs.

radumg commented:
Thanks @mjkkirschner for the detailed feedback & review, will get working on that option asap 👍 
I'll open a new PR for it as I tried re-basing this one and it's too messy.

FYI, I had a [working version](https://github.com/radumg/Dynamo/commit/bf1fe440706b63e156a44e7e20ceb1abaee46181) of the static `PreferenceSettings` property on the `DynamoModel` too, in case you were wondering what the consequences of that option would've been.

Quick question : when are you aiming to finish round of PR merges for next release ? I would like to get this one (should be ready this weekend) in there but also the next one which will look at support for `DYN` templates too.

radumg commented:
created a new PR #8122 based on `master` branch, please see that for the implementation of what was discussed above.
